### DISCUSSION TITLE
Running a large scale frictional interaction simulation with SOFA

### DISCUSSION THREAD

**Discussion by harshjn** [sofa-framework/sofa#4286] [Write / Run a simulation] [answered]
Hi, 
I am trying to reproduce my experimental results from my preprint: https://arxiv.org/abs/2310.01115
Basically, for a network of hyperelastic rods, as frictional interaction is enabled between the rods by compressing them tightly, the configuration after indentation will be different from the configuration before indentation as examplified in this **video results**:
**Low friction: No change in mesh configuration after indentation** https://1drv.ms/v/s!AhO10LFsAPgjlMVV_ON2gPI4VsIjJQ
**High friction: Change in mesh configuration after indentation** https://1drv.ms/v/s!AhO10LFsAPgjlMVWsU1gy4Gp3uaUZQ

The **codes can be found** in the scene files at
https://github.com/harshjn/MechanicalMemorySimulation/tree/main/Oct18

The scene file **FrictionMemory.scn** works, as shown in the video results above, but it crashes a lot due to failed convergence when collision elements get close.  However, **My main issue is** that, when I use large meshes as done in **FrictionMemoryLarge2.scn**, the simulation does not converge at all and I get the following error from LCPConstraint: No covergence, error = -nan after 9500 iterations.

I have made the simulation parallel and optimised the DefaultContactManager as well as LCPConstraint as shown in the following simulation setup lines in xml:
```xml
<Node name="root" gravity="0 0 0" dt="0.001">
<VisualStyle displayFlags="showCollisionModels" />
<VisualStyle displayFlags="showVisual showBehaviorModels" />
<LCPConstraintSolver maxIt="19500" tolerance="1e-8" initial_guess="true" build_lcp="false" multi_grid="false" printLog="0" mu="80"/>
<FreeMotionAnimationLoop parallelCollisionDetectionAndFreeMotion="true" parallelODESolving="true" /> 
<CollisionPipeline depth="30" draw="0" verbose="0" name="CollisionPipeline" />
<ParallelBruteForceBroadPhase/>
<ParallelBVHNarrowPhase/>
<LocalMinDistance name="Proximity" alarmDistance="0.5" contactDistance="0.25" useLMDFilters="0" />
<DefaultContactManager name="Response" response="FrictionContactConstraint" responseParams="mu=80"   />
 <EulerImplicitSolver name="cg_odesolver" printLog="false"  rayleighStiffness="0.15" rayleighMass="0.15" />
 <CGLinearSolver iterations="25" name="linear solver" tolerance="1.0e-15" threshold="1.0e-15" />
 ```

Please help @hugtalbot @bakpaul

**Comment by bakpaul** [comment, marked as answer]
Hi Harsh

First of all, your simulation are very cool, congrats !

I haven't had the time to test your scene (`FrictionMemoryLarge2.scn,`), but there is one thing that stroke me while reading it and it is that you are using an `UncoupledConstraintCorrection`. I don't know if you totally get what this means :it means that at the resolution step, each constraint will be treated independently. Meaning that the coupling between the constraints will not be taken into account during the contraint resolution stage. This has two advantages: 
1. The compliance computation is very quick, there is only a diagonal matrix generated without any computation expense
2. The constraint solving is quick because only the inversion of a diagonal matrix is necessary.

But it also has big drawbacks : 
1. Because you don't take the contraint coupling into account, you will never be able to compute the result of two antagonists constraints. This means that two constraints that are close will never be solved correctly and will lead to instability 
2. Because the constraint solving is done without any a priori knowledge of the mechanics (indeed you use a manually engineered matrix to couple them), the resulting motion of the object might be not physically realistic.

You talked about instability and crashed, I suspect that this is happening because you have a simulation with very close constraints and thus coupled, you should treat them accordingly. You need to use a solver that uses the right coupling between your constraints. There is two solution for that (each one leading to drop in FPS...) : 
1. `LinearSolverConstraintCorrection`: This is the most precise one that exist in SOFA. The downside is that this solver requires the computaiton of the compliance matrix at each times step (see [this page of the doc](https://www.sofa-framework.org/community/doc/simulation-principles/constraint/lagrange-constraint/) for more information), which take a lot of time. It can be optimized thank's to parallel implementations of the `SparseLDLSolver` that exist in the sofa plugin `MultiThreading` or using `ConstantSparsityPatternSystem` as your matrix system. 
2. `PrecomputedConstraintCorrection`: This one will be less precise, the idea is that, for the computation of the compliance, the inverse of the system matrix is needed. This component computes it only once at the begining and uses the same during the whole simulation. It has the advantage to have a mechanically sound coupling but may produce instability  with big rotation of your model. Which doesn't seem to be the case here.  

It has to be noted that using the first solution you'll need a direct solver as linear solver (such as the 
`SparseLDLSolver`) and one of such in every object node. For instance here you'll need one in the sponge node and one in the compressor node, and you'll need to remove the CGSolver. I am not sure about the ode solver, but usually I also add one per node and remove the one you put. 

Here is an example of a node using the first combination (linear solver + multithread + constant sparsity):
```xml
    <Node name="S" >
        <EulerImplicitSolver firstOrder="false" rayleighMass="0.1" rayleighStiffness="0.1" />
        <ConstantSparsityPatternSystem template="CompressedRowSparseMatrixd" name="A" printLog="false"/>
        <SparseLDLSolver name="precond" template="CompressedRowSparseMatrixd" parallelInverseProduct="true" linearSystem="@A"/>

        <MeshVTKLoader name="loader" filename="../Data/S.vtk"  />
        <TetrahedronSetTopologyContainer name="Container" src="@loader" />
        <TetrahedronSetTopologyModifier name="Modifier"/>
        <MechanicalObject name="mstate" template="Vec3d" />
        <TetrahedronFEMForceField name="FEM"  youngModulus="200" poissonRatio="0.45" method="large" />

        <MeshMatrixMass totalMass="0.15" />
        <Node name="Collision">
            <TriangleSetTopologyContainer name="Container" />
            <TriangleSetTopologyModifier name="Modifier"/>
            <Tetra2TriangleTopologicalMapping input="@../Container" output="@Container" />

            <MechanicalObject name="dofs" rest_position="@../mstate.rest_position"  />
            <TriangleCollisionModel name="Surface" proximity="0.02" />
            <IdentityMapping name="SurfaceMapping"/>
        </Node>
        <LinearSolverConstraintCorrection linearSolver="@precond"/>
    </Node>
```
I hope this helped you !
